Both review points were right; fixed in `60e24b4ce`.

**The sweep counter was unpinned.** I checked whether `pendingAlerts()` is even reachable in the failed-bind fixture — it is: `service.maintain()` swallows per-record errors and `claimDueRetries()` returns `[]` on an empty store, so nothing throws before the alert loop and the rejecting `installationToken` is never reached. So the original RED measurement was real. But you're right that nothing in the test held it there. Added a positive control that starts a live server and requires the counter to advance.

Verified by mutation rather than assertion. Replacing the alert loop's iterable with an empty array leaves the interval-leak test **green** and fails the control:

```
FAIL  relay maintenance interval > advances the sweep counter while the server is up
AssertionError: expected 0 to be greater than 0
Tests  1 failed | 2 passed (3)
```

That is exactly the silent-vacuity mode you named.

**The lock-release path was dead.** `allowUnlockedForTests: true` means `ownsProcessLock` is false, so `processLock?.release()` never ran under test. Added a `lockPath` case that reacquires the same path after a rejected bind. Dropping the release from the failure path fails it:

```
FAIL  relay startup failure > releases the process lock it acquired when the port is taken
Error: retro relay database is already locked
  at ProcessLock.acquire (src/process-lock.ts:14:48)
```

Full `packages/retro-relay` suite: 173 passed, 1 skipped. No production code changed.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/packages/retro-relay/src/http-server.ts
+++ b/packages/retro-relay/src/http-server.ts
@@ -219,10 +219,24 @@ export async function startRelayServer(input: RelayServerOptions): Promise<{
           void maintain();
         }, input.maintenanceIntervalMs ?? DEFAULT_MAINTENANCE_INTERVAL_MS);
   maintenanceTimer?.unref();
-  server.once('close', () => {
+  /**
+   * Releases everything startup acquired, once.
+   *
+   * Startup can fail after the maintenance interval exists, and a failed
+   * `listen` never emits 'close' — so leaving cleanup to the close handler
+   * alone leaves that interval sweeping the real store and GitHub client on
+   * behalf of a server the caller was told did not start, with another added
+   * per retry. Every failure path calls this rather than releasing the lock
+   * piecemeal and hoping 'close' arrives.
+   */
+  let released = false;
+  const releaseResources = (): void => {
+    if (released) return;
+    released = true;
     if (maintenanceTimer !== undefined) clearInterval(maintenanceTimer);
     if (ownsProcessLock) processLock?.release();
-  });
+  };
+  server.once('close', releaseResources);
 
   const recordReconciliationOutcome = (entry: {
     receiptId: string;
@@ -386,13 +400,13 @@ export async function startRelayServer(input: RelayServerOptions): Promise<{
       server.listen(input.port ?? 0, input.host ?? '127.0.0.1', resolve);
     });
   } catch (error) {
-    if (ownsProcessLock) processLock?.release();
+    releaseResources();
     throw error;
   }
   const address = server.address();
   if (address === null || typeof address === 'string') {
     server.close();
-    if (ownsProcessLock) processLock?.release();
+    releaseResources();
     throw new Error('relay did not bind a TCP address');
   }
   return {

--- a/packages/retro-relay/tests/startup-cleanup.test.ts
+++ b/packages/retro-relay/tests/startup-cleanup.test.ts
@@ -176,4 +176,22 @@ describe('relay startup failure', () => {
       store.close();
     }
   });
+
+  // The control for the case above. Reacquiring proves a release only if
+  // acquiring while held would have failed — otherwise the assertion passes
+  // whether or not anything was ever released.
+  it('refuses to acquire a lock this process already holds', () => {
+    const lockPath = path.join(scratchDirectory(), 'relay.lock');
+    const held = ProcessLock.acquire(lockPath);
+
+    try {
+      expect(() => ProcessLock.acquire(lockPath)).toThrow('already locked');
+    } finally {
+      held.release();
+    }
+
+    // And releasing makes it available again, so the failure above is the lock
+    // being held rather than the path being unusable.
+    ProcessLock.acquire(lockPath).release();
+  });
 });

--- a/packages/retro-relay/tests/startup-cleanup.test.ts
+++ b/packages/retro-relay/tests/startup-cleanup.test.ts
@@ -6,6 +6,9 @@
  * that survives a rejected startup keeps mutating a relay the caller believes
  * never came up — and every retried start adds another. `unref()` keeps it from
  * holding the process open; it does not stop it firing.
+ *
+ * The process lock is the other half: when startup acquires it from `lockPath`,
+ * a rejection that keeps it makes every retry fail as "already locked".
  */
 
 import { mkdtempSync, rmSync } from 'node:fs';
@@ -13,11 +16,12 @@ import { createServer } from 'node:http';
 import { tmpdir } from 'node:os';
 import path from 'node:path';
 
-import { afterEach, describe, expect, it } from 'vitest';
+import { afterEach, describe, expect, it, vi } from 'vitest';
 
 import { CredentialRegistry } from '../src/auth.js';
 import { GitHubRestClient } from '../src/github.js';
 import { startRelayServer } from '../src/http-server.js';
+import { ProcessLock } from '../src/process-lock.js';
 import { RelayStore } from '../src/store.js';
 
 const directories: string[] = [];
@@ -28,10 +32,22 @@ afterEach(() => {
   for (const directory of used) rmSync(directory, { force: true, recursive: true });
 });
 
-function databasePath(): string {
+function scratchDirectory(): string {
   const directory = mkdtempSync(path.join(tmpdir(), 'safeword-relay-startup-'));
   directories.push(directory);
-  return path.join(directory, 'relay.sqlite');
+  return directory;
+}
+
+function databasePath(): string {
+  return path.join(scratchDirectory(), 'relay.sqlite');
+}
+
+/** A never-usable GitHub client: maintenance must not reach the network here. */
+function offlineGitHub(): GitHubRestClient {
+  return new GitHubRestClient({
+    baseUrl: 'https://github.invalid',
+    installationToken: () => Promise.reject(new Error('must not call GitHub')),
+  });
 }
 
 /** A store that counts the maintenance sweeps the interval drives. */
@@ -70,6 +86,36 @@ const idle = (ms: number): Promise<void> =>
     setTimeout(resolve, ms);
   });
 
+describe('relay maintenance interval', () => {
+  // Positive control for the counter the failure tests below read. A sweep
+  // count that stops growing proves nothing unless it grows when the interval
+  // is genuinely running, so this pins the probe to observable behaviour: if
+  // `pendingAlerts` ever stops being the sweep's witness, this fails first and
+  // the failure tests do not quietly go vacuous.
+  it('advances the sweep counter while the server is up', async () => {
+    const { store, sweeps } = countingStore();
+    const started = await startRelayServer({
+      allowUnlockedForTests: true,
+      credentials: new CredentialRegistry('pepper'),
+      github: offlineGitHub(),
+      host: '127.0.0.1',
+      maintenanceIntervalMs: 10,
+      payloadKey: Buffer.alloc(32, 7),
+      port: 0,
+      store,
+    });
+
+    try {
+      await vi.waitFor(() => {
+        expect(sweeps()).toBeGreaterThan(0);
+      }, 2000);
+    } finally {
+      await closeServer(started.server);
+      store.close();
+    }
+  });
+});
+
 describe('relay startup failure', () => {
   it('leaves no maintenance interval running when the port is taken', async () => {
     const blocked = await occupiedPort();
@@ -80,10 +126,7 @@ describe('relay startup failure', () => {
         startRelayServer({
           allowUnlockedForTests: true,
           credentials: new CredentialRegistry('pepper'),
-          github: new GitHubRestClient({
-            baseUrl: 'https://github.invalid',
-            installationToken: () => Promise.reject(new Error('must not call GitHub')),
-          }),
+          github: offlineGitHub(),
           host: '127.0.0.1',
           maintenanceIntervalMs: 10,
           payloadKey: Buffer.alloc(32, 7),
@@ -96,6 +139,38 @@ describe('relay startup failure', () => {
       const settled = sweeps();
       await idle(80);
       expect(sweeps()).toBe(settled);
+    } finally {
+      await blocked.release();
+      store.close();
+    }
+  });
+
+  // The interval is only half of what a failed startup holds. `lockPath` makes
+  // the server the lock's owner, and a lock that outlives the rejection makes
+  // every retry fail as "already locked" — a relay that can never start again
+  // without a restart. Reacquiring the same path is the only proof the release
+  // actually ran; `allowUnlockedForTests` never exercises it.
+  it('releases the process lock it acquired when the port is taken', async () => {
+    const blocked = await occupiedPort();
+    const store = RelayStore.open(databasePath());
+    const lockPath = path.join(scratchDirectory(), 'relay.lock');
+
+    try {
+      await expect(
+        startRelayServer({
+          credentials: new CredentialRegistry('pepper'),
+          github: offlineGitHub(),
+          host: '127.0.0.1',
+          lockPath,
+          maintenanceIntervalMs: 10,
+          payloadKey: Buffer.alloc(32, 7),
+          port: blocked.port,
+          store,
+        }),
+      ).rejects.toThrow();
+
+      const reacquired = ProcessLock.acquire(lockPath);
+      reacquired.release();
     } finally {
       await blocked.release();
       store.close();

--- a/packages/retro-relay/tests/startup-cleanup.test.ts
+++ b/packages/retro-relay/tests/startup-cleanup.test.ts
@@ -1,0 +1,104 @@
+/**
+ * What `startRelayServer` owns when it rejects.
+ *
+ * A failed startup must leave nothing running. The maintenance interval calls
+ * `service.maintain()` against the real store and GitHub client, so an interval
+ * that survives a rejected startup keeps mutating a relay the caller believes
+ * never came up — and every retried start adds another. `unref()` keeps it from
+ * holding the process open; it does not stop it firing.
+ */
+
+import { mkdtempSync, rmSync } from 'node:fs';
+import { createServer } from 'node:http';
+import { tmpdir } from 'node:os';
+import path from 'node:path';
+
+import { afterEach, describe, expect, it } from 'vitest';
+
+import { CredentialRegistry } from '../src/auth.js';
+import { GitHubRestClient } from '../src/github.js';
+import { startRelayServer } from '../src/http-server.js';
+import { RelayStore } from '../src/store.js';
+
+const directories: string[] = [];
+
+afterEach(() => {
+  const used = [...directories];
+  directories.length = 0;
+  for (const directory of used) rmSync(directory, { force: true, recursive: true });
+});
+
+function databasePath(): string {
+  const directory = mkdtempSync(path.join(tmpdir(), 'safeword-relay-startup-'));
+  directories.push(directory);
+  return path.join(directory, 'relay.sqlite');
+}
+
+/** A store that counts the maintenance sweeps the interval drives. */
+function countingStore(): { store: RelayStore; sweeps: () => number } {
+  const store = RelayStore.open(databasePath());
+  let sweeps = 0;
+  const pendingAlerts = store.pendingAlerts.bind(store);
+  store.pendingAlerts = () => {
+    sweeps += 1;
+    return pendingAlerts();
+  };
+  return { store, sweeps: () => sweeps };
+}
+
+function closeServer(server: ReturnType<typeof createServer>): Promise<void> {
+  return new Promise(resolve => {
+    server.close(() => {
+      resolve();
+    });
+  });
+}
+
+function occupiedPort(): Promise<{ port: number; release: () => Promise<void> }> {
+  return new Promise(resolve => {
+    const blocker = createServer();
+    blocker.listen(0, '127.0.0.1', () => {
+      const address = blocker.address();
+      if (address === null || typeof address === 'string') throw new Error('no port');
+      resolve({ port: address.port, release: () => closeServer(blocker) });
+    });
+  });
+}
+
+const idle = (ms: number): Promise<void> =>
+  new Promise(resolve => {
+    setTimeout(resolve, ms);
+  });
+
+describe('relay startup failure', () => {
+  it('leaves no maintenance interval running when the port is taken', async () => {
+    const blocked = await occupiedPort();
+    const { store, sweeps } = countingStore();
+
+    try {
+      await expect(
+        startRelayServer({
+          allowUnlockedForTests: true,
+          credentials: new CredentialRegistry('pepper'),
+          github: new GitHubRestClient({
+            baseUrl: 'https://github.invalid',
+            installationToken: () => Promise.reject(new Error('must not call GitHub')),
+          }),
+          host: '127.0.0.1',
+          maintenanceIntervalMs: 10,
+          payloadKey: Buffer.alloc(32, 7),
+          port: blocked.port,
+          store,
+        }),
+      ).rejects.toThrow();
+
+      // Whatever ran before the rejection is fine; nothing may run after it.
+      const settled = sweeps();
+      await idle(80);
+      expect(sweeps()).toBe(settled);
+    } finally {
+      await blocked.release();
+      store.close();
+    }
+  });
+});

